ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,225 @@ func (ps *prioritySampler) apply(spn *span) {
+	mu           sync.Mutex
+	ts           int64
+	allowed      int
+	total        int
+	previousRate float64
+}
+
+func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	rate := sampleRate()
+	return &rulesSampler{
+		rules:   samplingRules(rules),
+		rate:    rate,
+		limiter: newRateLimiter(rate),
+	}
+}
+
+// samplingRules validates the user-provided rules and returns an internal representation.
+// If the DD_TRACE_SAMPLING_RULES environment variable is set, then the rules from
+// tracer.WithSamplingRules are ignored.
+func samplingRules(rules []SamplingRule) []SamplingRule {
+	rulesFromEnv := os.Getenv("DD_TRACE_SAMPLING_RULES")
+	if rulesFromEnv != "" {
+		rules = rules[:0]
+		jsonRules := []struct {
+			Service   string      `json:"service"`
+			Operation string      `json:"operation"`
+			Rate      json.Number `json:"rate"`
+		}{}
+		err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+		if err != nil {
+			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+			return nil
+		}
+		for _, v := range jsonRules {
+			if v.Rate == "" {
+				log.Warn("error parsing rule: rate not provided")
+				continue
+			}
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Operation != "":
+				rules = append(rules, ServiceOperationRule(v.Service, v.Operation, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Operation != "":
+				rules = append(rules, OperationRule(v.Operation, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+func sampleRate() float64 {
+	rate := 0.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return rate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", rate, err)
+		return rate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", rate, r)
+	return rate
+}
+
+func newRateLimiter(r float64) *rate.Limiter {
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return nil
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return nil
+	}
+	switch {
+	case l < 0.0:
+		return nil
+	case l == 0.0:
+		return rate.NewLimiter(0.0, 0)
+	case (r * l) < 1.0:
+		return rate.NewLimiter(rate.Limit(l), 1)
+	default:
+		return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+	}
+}
+
+func (rs *rulesSampler) apply(span *span) bool {
+	matched := false
+	sr := 0.0
+	for _, v := range rs.rules {
+		if v.match(span) {
+			matched = true
+			sr = v.Rate
```

Comment:
`sr` is way too cryptic for a human. Can we use something that makes sense?

Reply:
Changed to `rate`